### PR TITLE
Fix #4796

### LINKED ISSUE #4796: Nice syntax for breakpoints on "module_base + offset" instead of "db @ `dm~library.so:0[2]`+ 0x50"
Since module bases can change per process execution, but offsets stay the same, support for breaking on a module_base + offset would be very helpful.

I picture something like "db (~stagefright.so) + 0x31337"...

The module base address should be resolved, offset added, and breakpoint inserted.

### PR DESCRIPTION
**Pull request by alvarofe**
WIP since there are overlaps with the name of the modules so probably we need to assign them an ID or something. Also i chose `dbmm` for this.

```
[0x5568d52c3df0]> dm
sys 120K 0x00005568d52be000 * 0x00005568d52dc000 s -r-x /usr/bin/ls /usr/bin/ls
sys   8K 0x00005568d54db000 - 0x00005568d54dd000 s -r-- /usr/bin/ls /usr/bin/ls
sys   4K 0x00005568d54dd000 - 0x00005568d54de000 s -rw- /usr/bin/ls /usr/bin/ls
sys   4K 0x00005568d54de000 - 0x00005568d54df000 s -rw- unk0 unk0
sys 1.6M 0x00007f194b9a5000 - 0x00007f194bb3d000 s -r-x /usr/lib/libc-2.23.so /usr/lib/libc-2.23.so
sys   2M 0x00007f194bb3d000 - 0x00007f194bd3d000 s ---- /usr/lib/libc-2.23.so /usr/lib/libc-2.23.so
sys  16K 0x00007f194bd3d000 - 0x00007f194bd41000 s -r-- /usr/lib/libc-2.23.so /usr/lib/libc-2.23.so
sys   8K 0x00007f194bd41000 - 0x00007f194bd43000 s -rw- /usr/lib/libc-2.23.so /usr/lib/libc-2.23.so
sys  16K 0x00007f194bd43000 - 0x00007f194bd47000 s -rw- unk1 unk1
sys  16K 0x00007f194bd47000 - 0x00007f194bd4b000 s -r-x /usr/lib/libcap.so.2.25 /usr/lib/libcap.so.2.25
sys   2M 0x00007f194bd4b000 - 0x00007f194bf4b000 s ---- /usr/lib/libcap.so.2.25 /usr/lib/libcap.so.2.25
sys   4K 0x00007f194bf4b000 - 0x00007f194bf4c000 s -r-- /usr/lib/libcap.so.2.25 /usr/lib/libcap.so.2.25
sys   4K 0x00007f194bf4c000 - 0x00007f194bf4d000 s -rw- /usr/lib/libcap.so.2.25 /usr/lib/libcap.so.2.25
sys 140K 0x00007f194bf4d000 - 0x00007f194bf70000 s -r-x /usr/lib/ld-2.23.so /usr/lib/ld-2.23.so
sys  16K 0x00007f194c165000 - 0x00007f194c169000 s -rw- unk2 unk2
sys   4K 0x00007f194c16f000 - 0x00007f194c170000 s -r-- /usr/lib/ld-2.23.so /usr/lib/ld-2.23.so
sys   4K 0x00007f194c170000 - 0x00007f194c171000 s -rw- /usr/lib/ld-2.23.so /usr/lib/ld-2.23.so
sys   4K 0x00007f194c171000 - 0x00007f194c172000 s -rw- unk3 unk3
sys 132K 0x00007ffee56c9000 - 0x00007ffee56ea000 s -rw- [stack] [stack]
sys   8K 0x00007ffee57fc000 - 0x00007ffee57fe000 s -r-- [vvar] [vvar]
sys   8K 0x00007ffee57fe000 - 0x00007ffee5800000 s -r-x [vdso] [vdso]
sys   4K 0xffffffffff600000 - 0xffffffffff601000 s -r-x [vsyscall] [vsyscall]
[0x5568d52c3df0]> dbmm /usr/lib/libcap.so.2.25 90
[0x5568d52c3df0]> db
0x7f194bd4705a - 0x7f194bd4705b 1 --x sw break enabled cmd="" name="139746623123546"
[0x5568d52c3df0]>
```

### CODE REVIEW

**Comment by radare** [comment]
i would go for dbm to accept an offset as argument and then with this offset determine the modulename, grab the base address and compute the delta.

the thing is that those breakpoints should be stored internally as module+offset, so when you reload the project or the process they keep being mesningfull

the problem is that not all maps are loaded always or even in the same order.so neither using index or name is a valid enough solution for all use cases

**Comment by radare** [comment]
Imho your `dbmm` should override the current `dbm` .. and we should find a new name for `dbm`.. maybe `dbb` for base

File: libr/core/cmd_debug.c [hunk 1]
```diff
@@ -1859,7 +1859,27 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
 		break;
 	case 'm': // "dbm"
 		if (input[2]) {
-			core->dbg->bp->delta = (st64)r_num_math (core->num, input+2);
+		    	if (input[2] == 'm' && input[3] != '\0' && input[4] != '\0') { //dbmm
+			    	char *string = strdup(input+4);
```
Comment (radare) [inline libr/core/cmd_debug.c]: missing spaces around + and before (

File: libr/core/cmd_debug.c [hunk 2]
```diff
@@ -1859,7 +1859,27 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
 		break;
 	case 'm': // "dbm"
 		if (input[2]) {
-			core->dbg->bp->delta = (st64)r_num_math (core->num, input+2);
+		    	if (input[2] == 'm' && input[3] != '\0' && input[4] != '\0') { //dbmm
```
Comment (radare) [inline libr/core/cmd_debug.c]: drop the unnecessary `!= '\0'`

File: libr/core/cmd_debug.c [hunk 3]
```diff
@@ -1859,7 +1859,27 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
...
+			    	char *string = strdup(input+4);
+				char *module = NULL;
```
Comment (radare) [inline libr/core/cmd_debug.c]: sort vars by strlen

File: libr/core/cmd_debug.c [hunk 4]
```diff
@@ -1859,7 +1859,27 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
...
+				ut64 offset = 0;
+				RList *list;
+				RListIter *iter;
+				RDebugMap *map;
+				module = strtok (string, " ");
+				offset = (ut64)r_num_math (core->num, strtok (NULL, ""));
+				list = r_debug_modules_list (core->dbg);
+				r_list_foreach (list, iter, map) {
+				    	if (!strncmp (map->name, module, R_MIN(strlen(map->name), strlen(module)))) {
```
Comment (radare) [inline libr/core/cmd_debug.c]: missing spaces before (

File: libr/core/cmd_debug.c [hunk 5]
```diff
@@ -1859,7 +1859,27 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
...
+					    	char cmd[128] = {0};
```
Comment (radare) [inline libr/core/cmd_debug.c]: why not just use r_core_cmdf

**Comment by alvarofe** [comment]
I will try to take care of it along this week ;)

**Comment by radare** [comment]
When you say this week you mean tonight?

**Comment by alvarofe** [comment]
i will try it but i don't promise nothing xD

File: libr/core/cmd_debug.c [hunk 6]
```diff
@@ -1857,12 +1857,35 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
...
 		break;
-	case 'm': // "dbm"
+	case 'b':	
 		if (input[2]) {
 			core->dbg->bp->delta = (st64)r_num_math (core->num, input+2);
 		} else {
 			r_cons_printf ("%"PFMT64d"\n", core->dbg->bp->delta);
 		}
+	case 'm': // "dbm"
+		if (input[2]) {
```
Comment (radare) [inline libr/core/cmd_debug.c]: merge those two in a single conditional

File: libr/core/cmd_debug.c [hunk 7]
```diff
@@ -1857,12 +1857,35 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
...
+		    	if (input[3]) { 
+			    	char *string = strdup(input + 3);
```
Comment (radare) [inline libr/core/cmd_debug.c]: missing space before (

**Comment by alvarofe** [comment]
What about extend `RBreakpoint` to store that info. We would save the module name, delta and if it's necessary more. Regarding with mapping to choose probably the solution would be something interactive

```
[0x7fff5fc01000]> dbm dyld 90
We have found 2 coincidences
sys 224K 0x00007fff5fc00000 * 0x00007fff5fc38000 s -r-x /usr/lib/dyld 05_copy/usr/lib/dyld-rwx
sys 012K 0x00007fff5fc38000 - 0x00007fff5fc3b000 s -rw- /usr/lib/dyld 06_copy/usr/lib/dyld-rwx
option: 2
```
Then maybe saving the size of the mapping or other data (kind of heuristics) we would be able to reload a project an set the breakpoints correctly or not save `module+offset` into the project file just only plain breakpoints

**Comment by radare** [comment]
Yep makes sense. But what if theres more than one region with the same name? Maybe module[ ctr ] + delta

**Comment by jduck** [comment]
A breakpoint should never (??) be in a section other than a text section. AFAIK those sections are always singular-per-module and contiguous. Please correct me if I'm wrong!

**Comment by jduck** [comment]
The way WinDbg does it is by using the basename of the library. AFAIK, if there are multiple loaded with the same name it appends a suffix based on load order. IF that doesn't work out (the libraries are really loaded in random order, which might happen on newer androids), we can always fall back to the old way of locating the section by some other heuristic (size, full path, or similar).

**Comment by radare** [comment]
JIT for example, only maps with exec perms. In fact we should check also if we set a hw write breakpoint on a readonly map.

But maybe the user explictly knows and wants to do that because maybe at some point the program changes the protections of this region if the os permits it.
Setting a bp on an unallocated area should not be permitted either. Unless its a hwbp

**Comment by radare** [comment]
ping

**Comment by jduck** [comment]
Very good point about JIT, though dealing with that would require a breakpoint that sets a breakpoint type of situation anyway. It's highly doubtful that any "break module+offset" would help at all in that case. There isn't a module, the mapping probably has no name, and it's address is not known except to the JIT engine/JIT consumer itself.

Good point about the hw write breakpoint on a read-only map, but it is possible that such a map could change.. I agree that letting the user do what they want makes sense here (hopefully they don't shoot themselves).

Agree that validation needs to happen if someone tries an unmapped area. The only way that could work is by implement "future breakpoints".

**Comment by radare** [comment]
Setting a breakpont on unmapped area is not really a problem because if its mapped at some point the bp will be set after continue or step

**Comment by radare** [comment]
And yes, r2 aims to be a low level debugger and such behaviours are expected to give more power to the user, between some constraints to kot make them unusable. Maybe a warning is enough

**Comment by jduck** [comment]
Good point!

**Comment by alvarofe** [comment]
Updated

https://asciinema.org/a/7z66zwdw1xxyy1zp74uilj1gw

I took the idea from jduck's branch to add `r_debug_bp_add` to simplify the API. I grab the base address from the modules list instead of mappings. Once i compute the final address i check whether it falls inside a valid mapping. If the address falls inside of mapping that doesn't have `x` a warning is printed but the bp is inserted anyway. 

`db*` now reflect `dbm` to save breakpoints info into project file though is not saved actually.

**Comment by radare** [comment]
lgtm, ok to merge for everyone else?

maybe it will be good to have a test too

**Comment by radare** [comment]
Thanks!

**Comment by jduck** [comment]
Nice. I've merged it into my debugger-fixes branch too.

**Pull request merged**